lib/sinon/test.js

```diff
@@ -55,6 +55,15 @@
                     sandbox.verifyAndRestore();
                 }
 
+                if (result && typeof result.then === "function" && typeof result.catch === "function") {
```

Comment:
If you simplify this, it will work with more Promise libraries like jQuery (before v3). If you only use `.then(onResolve, onReject)`. I do realize that old jQuery was not Promises/A+ compliant. It's just a thought and almost functionally the same.

The way this is currently written, it will actually catch the failure if something were to throw in the `onResolve` of the `then` which seems unnecessary. 

I apologize for the string of comments instead of all at once.